### ISSUE TITLE
flux exec vs wreckrun behavior forwarding environment variables

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#864]
Was messing around trying to get Magpie to work under Slurm & flux and got a very trivial case to work.

I noticed that the initial environment variables created by Slurm ```srun``` don't get "forwarded" to tasks executed by ```wreckrun``` but they do for ```exec```.

Here's what I mean.

```
$ cat do_some_stuff.sh
#!/bin/sh
hostname
env | grep SLURM_JOB_ID
env | grep FOO
exit 0

$ cat fluxexec.sh
#!/bin/sh
flux exec do_some_stuff.sh
exit 0

$ srun -N4 -n4 src/cmd/flux start fluxexec.sh
srun: Job is in held state, pending scheduler release
srun: job 1159741 queued and waiting for resources
srun: job 1159741 has been allocated resources
hype254
SLURM_JOB_ID=1159741
FOO=bar
hype256
hype255
SLURM_JOB_ID=1159741
SLURM_JOB_ID=1159741
FOO=bar
hype257
FOO=bar
SLURM_JOB_ID=1159741
FOO=bar
```

As expected, Slurm starts up flux, flux runs ```fluxexec.sh```, which runs ```do_some_stuff.sh``` on all nodes via ```flux exec```.  ```do_some_stuff.sh``` outputs the hostname then looks to see if environment variables ```SLURM_JOB_ID``` and ```FOO``` are there.  ```FOO``` is something I created locally before running.  As you can see in the output, ```SLURM_JOB_ID``` and ```FOO``` are forwarded everywhere and available on each node.

Here's the case w/ ```wreckrun```

```
$ cat fluxwreckrun.sh
#!/bin/sh
flux wreckrun -N4 do_some_stuff.sh
exit 0

$ srun -N4 -n4 src/cmd/flux start fluxwreckrun.sh
srun: Job is in held state, pending scheduler release
srun: job 1159742 queued and waiting for resources
srun: job 1159742 has been allocated resources
hype254
hype255
hype256
hype257
FOO=bar
FOO=bar
FOO=bar
FOO=bar
```

The SLURM environment variable ```SLURM_JOB_ID``` doesn't get "forwarded" but ```FOO``` does.

Is this intended in some way?  Or am I not understanding something?  The SLURM environment variables were missing in the ```kvs``` too, but ```FOO``` was there.

I use the environment variables b/c I use to the jobname and jobid as unique identifiers for creation of various info and files in Magpie.  Obviously in the future with a bootstrapped flux these would perhaps be available in other ways, but I believe it's what I have to use at the moment.  Here's the section of a Magpie script I updated for flux, with some "HARDCODED".

```bash
export MAGPIE_CLUSTER_NODERANK="${FLUX_NODE_ID}"
export MAGPIE_NODE_COUNT="${FLUX_JOB_NNODES}"
MAGPIE_NODELIST=""
for i in `seq 0 $((${FLUX_JOB_NNODES} - 1))`
do
  fluxhostname=`flux kvs get resource.hwloc.xml.${i} | grep "HostName" | sed -e 's/<info name="HostName" value="\(.*\)"\/>/\1/' | tr -d '[[:space:]]'`
  MAGPIE_NODELIST=${MAGPIE_NODELIST}${MAGPIE_NODELIST:+","}${fluxhostname}
done
export MAGPIE_NODELIST
# HARDCODED
export MAGPIE_JOB_NAME="testingflux"
# HARDCODED
export MAGPIE_JOB_ID="1234"
```

As an aside, that's some hackery I have to do in the middle there to get the nodes I'm running on.  I can add it as a todo to create a FLUX_NODELIST environment variable or some equivalent.

**Comment by SteVwonder** [comment]
I believe this was intentional due to some issues with MPI.  IIRC, MPI applications would wire up in the SLURM allocation instead of the Flux allocation upon detection of certain SLURM environment variables (not 100% sure though).  The wreck code that is sanitizing the environment is https://github.com/flux-framework/flux-core/blob/18ae69704b5b2e2ee7f221e336bf6f253fc07607/src/bindings/lua/wreck.lua#L133

**Comment by grondo** [comment]
Thanks @SteVwonder, I was just typing up a response but you beat me to it.

This is correct, we don't want to export SLURM variables to MPI apps as they may think they're running under Slurm and attempt to connect back to srun in some cases, or use the Slurm PMI api.

Environment export on the `wrexecd` side is handled by a plugin `src/modules/wreck/lua.d/01-env.lua`, which reads common variables per-instance from top-level `lwj.` directory, then per-job, per-task, etc.

Can you use FLUX_JOB_ID instead of hard-coded MAGPIE_JOB_ID?

**Comment by chu11** [comment]
@SteVwonder Ahh thanks, I think I grepped only on *.[ch] files which is why I did come upon that.

**Comment by grondo** [comment]
Apologies about the current state of the `wreck` prototype as well. It is definitely not easy to get a list of hosts or even ranks on which a wreckrun job is running or has been scheduled. Your kludge above unfortunately won't work in general because it is not guaranteed that a job will be running on the ranks `0-${FLUX_JOB_NNODES}-1`

We'll need a better way to get a list of ranks/nodes for a job (eventually this will be easily handled by the jobspec)

**Comment by chu11** [comment]
@garlick Yeah, I think that's closer to what I'm thinking of.  I suppose a similar ```session-name``` would be useful.  Although ```session-id``` seems to come out of PMI.  There doesn't seem to be a similar "session-name" in PMI?

I'm thinking what I should do is just stuff whatever information I want (e.g. SLURM_JOB_NAME, SLURM_NODELIST, etc.) into the KVS.

**Comment by grondo** [comment]
> My kludge above is definitely not safe in the general case, but I think it's safe if --ntasks=#NODES and --tasks-per-node=1?

Yes, you are correct. Sorry I didn't have the whole picture.

**Comment by grondo** [comment]
> I'm thinking what I should do is just stuff whatever information I want (e.g. SLURM_JOB_NAME, SLURM_NODELIST, etc.) into the KVS.

I may not be remembering correctly, but Slurm "job name" is meant to be set by the user on the command line, and is not guaranteed to be unique. I think by default in Slurm it is set to `argv[0]` or something if not already set.

Copying SLURM env vars you want to save into the kvs is not a bad idea. Or you could just do `$ export FLUX_JOB_NAME=$SLURM_JOB_NAME` in the initial program shell.

**Comment by garlick** [comment]
What's the job name used for?  Is it so the "enclosing instance", e.g. slurm or flux, can list the job as "magpie" in `sinfo` or `flux wreck ls`?

**Comment by grondo** [comment]
I'll also mention that there is a way to set a default job environment for *all* jobs within a session without setting environment variables in the shell. However, the interface is a bit klunky  (there isn't one). If the wrexec `01-env.lua` plugin finds a kvs entry `lwj.environ` it will treat it as a json dictionary of environment variables to set/override in every job's environment, before applying per-job and per-task environment.

**Comment by grondo** [comment]
I pulled off the FLUX_NODELIST request into a separate issue #866

**Comment by chu11** [comment]
> I may not be remembering correctly, but Slurm "job name" is meant to be set by the user on the command line, and is not guaranteed to be unique. I think by default in Slurm it is set to argv[0] or something if not already set.

@grondo Yeah, slurm job name isn't unique, it's really the job id that I need that is unique.

> What's the job name used for? Is it so the "enclosing instance", e.g. slurm or flux, can list the job as "magpie" in sinfo or flux wreck ls?

@garlick The job name is mostly for organizational purposes, so that's why hardcoding some random string was ok for the time being.  But now that you mention it, the listing into ```squeue``` is convenient too.

**Comment by chu11** [comment]
I think everything here got answered and Mark created #866 for the remaining issue.  So I'll close.

**Comment by grondo** [comment]
In Slurm, I think job name is meant to be set by the user and is not a
unique generated name. However, your could pass the Slurm job name down to
the flux instance via something like `-o,-Sjob-name=$SLURM_JOB_NAME`, or
set your own name similarly... Maybe this is what you already suggested
@garlick?

**Comment by chu11** [comment]
That's more or less what I was thinking, but thought about using the kvs instead.  Is there any advantage/disadvantage to use the broker attrributes instead of kvs?

**Comment by garlick** [comment]
>  Is there any advantage/disadvantage to use the broker attrributes instead of kvs?

The KVS is global, and you may have to bend over backwards to load initial content since there's not currently a way to initialize a key on the command line.  The attributes are local (per broker rank) and can be initialized from the command line with -S as @grondo described, which incidentally would set that value globally because the -o command line options are passed to all broker invocations.

In short, I think the attributes are better here.
